src/cdk/portal/portal.ts

```diff
@@ -103,42 +104,38 @@ export class ComponentPortal<T> extends Portal<ComponentRef<T>> {
   }
 }
 
-
 /**
  * A `TemplatePortal` is a portal that represents some embedded template (TemplateRef).
  */
-export class TemplatePortal extends Portal<Map<string, any>> {
+export class TemplatePortal<C> extends Portal<C> {
   /** The embedded template that will be used to instantiate an embedded View in the host. */
-  templateRef: TemplateRef<any>;
+  templateRef: TemplateRef<C>;
 
   /** Reference to the ViewContainer into which the template will be stamped out. */
   viewContainerRef: ViewContainerRef;
 
-  /**
-   * Additional locals for the instantiated embedded view.
-   * These locals can be seen as "exports" for the template, such as how ngFor has
-   * index / event / odd.
-   * See https://angular.io/docs/ts/latest/api/core/EmbeddedViewRef-class.html
-   */
-  locals: Map<string, any> = new Map<string, any>();
+  context: C | undefined;
 
-  constructor(template: TemplateRef<any>, viewContainerRef: ViewContainerRef) {
+  constructor(template: TemplateRef<any>, viewContainerRef: ViewContainerRef, context?: C) {
     super();
     this.templateRef = template;
     this.viewContainerRef = viewContainerRef;
+    if (context) {
+     this.context = context;
+    }
   }
 
   get origin(): ElementRef {
     return this.templateRef.elementRef;
   }
 
-  attach(host: PortalHost, locals?: Map<string, any>): Map<string, any> {
-    this.locals = locals == null ? new Map<string, any>() : locals;
+  attach(host: PortalHost, context?: C): C {
+    this.context = context!;
```

Comment:
Indeed.

I think `this.context = context || this.context;` is more readable but I used your approach

It requires a slight modification to be friends with typescript:
```
attach(host: PortalHost, context: C | undefined = this.context): C {
```